FOUNDATIONS.md

```diff
@@ -0,0 +1,160 @@
+# IPLD Foundational Principles
+
+This document outlines parts of IPLD that should and should not be changed to
+ensure the success of future improvements (especially type systems).
+
+**Definitions**
+
+* Block: A block is a chunk of an IPLD DAG, encoded in a format. Blocks have CIDs.
+* Fragment: A piece of an IPLD DAG. Blocks contain fragments.
+* Node: A node is a *point* in an IPLD DAG (object, array, number, etc.).
+* Link: A link is an IPLD Node that points to another IPLD Node.
+* Path: A paths a human readable pointer to an IPLD Node.
+
+## Linked
+
+IPLD must support linking to any IPLD node (even if the node is in the middle of
+a block). That is, IPLD must support arbitrary IPLD paths in links.
+
+**Motivation:** Considering this in the context of programming languages, not
+being able to *store* a pointer to a struct *inside* of another struct would be
+severely limiting.
+
+NOTE: We don't currently support arbitrary paths but, in the context of
+programming, we really need to.
+
+## Immutable
+
+IPLD links must be immutable. We'll likely define a mutable link spec on top of
+IPLD but there needs to be an immutable layer at the bottom.
+
+**Motivation:** *Having* an immutable layer is important for a lot of analysis,
+memoization, type checking, etc.
+
+## Multicodecs Are Not Types
+
+It's impossible to understand IPLD data at a *structural* level if we don't know
+the format. Therefore, we should avoid introducing new formats unnecessarily as
+*every* IPLD implementation needs to support these new formats.
+
+## No Non-Local Reasoning
+
+An IPLD block should never be interpreted in the context of *anything* not
+contained in the block (and CID).
+
+For example, assuming we add support for relative links, the following
+definition of `foo` would not be a valid IPLD block:
+
+```
+var foo = {
+  // points outside of the current block, into the parent's "baz" field.
+  "baz": {"/": "../../baz"}
+}
+var bar = {
+  "foo": CidOf(foo),
+  // `/foo/baz` points here.
+  "baz": "something"
+}
+
+// resolution throug block `foo` depends on block `bar`.
+Resolve("/ipld/${CidOf(bar)}/foo/baz/")
+```
+
+For the same reason, IPLD links can't rely on an authority (e.g., a blockchain).
+
+Note: Links like this can still be encoded at the application level but they
+won't be handled by the IPLD resolver (and won't get the special "link" type).
+
+**Motivation:** IPLD needs to be easy to reason about.
+
+**Caveat:**
+
+We *may* want to relax this if we want to move schemas into separate,
+deduplicated blocks (referenced by CID). If we do that, we'd need to fetch a
+block's schema before being able to interpret the it.
```

Comment:
I don’t actually what the contradiction here is.

For one thing, if a block *without* a schema can be decoded at the block level then it’s always valid and understandable. The fact that schemas greatly *improve* the interpretation should not invalidate the previous statements.

I do think you may want to iterate a bit on the above statements regarding context. Data *inside* of a block is atomic and does not *rely* on outside context, but outside context can still be **applied** to the data within a block. Think about an file encoded in IPFS, you could look at the list of links to raw blocks and they contain a lot of atomic information on their own, they are an ordered list of raw binary chunks, but the fact that they represent a file isn’t contained *in* that data, the only context in which you can interpret them as parts of a file is from having read the file metadata which points to this list of raw binary chunks.

Like so much with DAGs, context can only be applied in one direction ;)

Reply:
Yeah, I tried to tear into some of that in https://github.com/ipld/specs/pull/156 .  Does that read better?

I like this example about files as well, maybe we should roll that into the diff as well.

> For one thing, if a block *without* a schema can be decoded at the block level then it’s always valid and understandable. The fact that schemas greatly *improve* the interpretation should not invalidate the previous statements.

:heavy_plus_sign: :100: 

> Like so much with DAGs, context can only be applied in one direction ;)

Super good statement.  I've been trying to come up with a way to phrase that for a while now, and I like this statement of it.

Some of the other negative examples in this vicinity use examples of cyclic links, and while we can all agree *those* are problematic, it's no argument against accumulating understanding as we follow *down* a DAG!

Reply:
Ah, yes, the new PR looks great. Sorry, my email backlog is in chronological order ;)

Reply:
:heart: